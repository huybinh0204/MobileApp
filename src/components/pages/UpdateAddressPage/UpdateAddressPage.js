import { yupResolver } from '@hookform/resolvers/yup';
import { useNavigation, useRoute } from '@react-navigation/native';
import React from 'react';
import { useForm } from 'react-hook-form';
import { object, string } from 'yup';
import { Form, FormInput, FormSubmitButton } from '_components/atoms';
import { SecondaryScreenLayout } from '_components/organisms';
import { NAVIGATION, REGEX } from '_constants';
import strings from '_localization/index';
import { ButtonContainer, Container } from './UpdateAddressPage.styles';

const schema = object().shape({
  addressLine1: string()
    .required(strings.signUp.requiredField)
    .matches(REGEX.address, strings.signUp.mainAddressError),
  addressLine2: string().matches(REGEX.alphanumeric, strings.signUp.alphanumericError),
});

const UpdateAddressPage = () => {
  const { params } = useRoute();
  const { navigate } = useNavigation();

  const defaultValues = {
    addressLine1: params?.addressLine1 ?? '',
    addressLine2: params?.addressLine2 ?? '',
  };

  const formMethods = useForm({ defaultValues, resolver: yupResolver(schema) });
  const line1 = formMethods.watch('addressLine1');
  const line2 = formMethods.watch('addressLine2');
  const isButtonDisabled = line1 === '';

  const handleSubmit = ({ addressLine1, addressLine2 }) => {
    navigate(NAVIGATION.settings.updateCity, {
      ...params,
      addressLine1,
      addressLine2,
      changed: line1 !== defaultValues.addressLine1 || line2 !== defaultValues.addressLine2,
    });
  };

  return (
    <SecondaryScreenLayout testID="UpdateAddressPage" title={strings.signUp.address.header}>
      <Container>
        <Form formMethods={formMethods}>
          <FormInput
            autoComplete="street-address"
            name="addressLine1"
            placeholder={strings.signUp.address.line1PlaceHolder}
            testID="addressLine1"
            textContentType="streetAddressLine1"
          />
          <FormInput
            name="addressLine2"
            placeholder={strings.signUp.address.line2PlaceHolder}
            testID="addressLine2"
            textContentType="streetAddressLine2"
          />
          <ButtonContainer>
            <FormSubmitButton
              disabled={isButtonDisabled}
              onSubmit={handleSubmit}
              testID="signUpNameSubmitButton"
            >
              {strings.signUp.continue}
            </FormSubmitButton>
          </ButtonContainer>
        </Form>
      </Container>
    </SecondaryScreenLayout>
  );
};

export default UpdateAddressPage;
