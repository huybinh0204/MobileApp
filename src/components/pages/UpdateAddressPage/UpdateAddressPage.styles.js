import styled from '@emotion/native';

export const Container = styled.View(({ theme }) => ({
  flex: 1,
  padding: theme.spacing.m,
}));

export const ButtonContainer = styled.View({
  flex: 1,
  alignItems: 'center',
  justifyContent: 'flex-end',
});
