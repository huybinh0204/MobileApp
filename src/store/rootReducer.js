import { combineReducers } from 'redux';
import actionWatchReducer from 'redux-action-watch/lib/reducer';
import { resettableReducer } from 'reduxsauce';
import { AccountReducer } from './account';
import { AuthenticationReducer } from './authentication';
import { CustomerReducer } from './customer';
import { FundDirectDepositStoreReducer } from './pageStore/fundDirectDepositStore';
import { RegisterReducer } from './register';
import { TransferReducer } from './transfer';

const resettable = resettableReducer('RESET');

export default combineReducers({
  account: resettable(AccountReducer),
  auth: resettable(AuthenticationReducer),
  customer: resettable(CustomerReducer),
  fundDirectDepositStoreReducer: resettable(FundDirectDepositStoreReducer),
  register: RegisterReducer,
  transfer: resettable(TransferReducer),
  watcher: actionWatchReducer, // Must be last
});
