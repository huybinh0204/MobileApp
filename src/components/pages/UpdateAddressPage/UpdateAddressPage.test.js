import { useNavigation } from '@react-navigation/native';
import { fireEvent, render, waitFor } from '@testing-library/react-native';
import React from 'react';
import { createStore } from 'redux';
import { NAVIGATION } from '_constants';
import strings from '_localization/index';
import rootReducer from '_store/rootReducer';
import ComponentWithProviders from '_test/mocks/ComponentWithProviders';
import { navigationMock } from '_test/mocks/navigation';
import UpdateAddressPage from './UpdateAddressPage';

jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useNavigation: jest.fn(),
}));

describe('UpdateAddressPage', () => {
  let mockStore;
  const mockNavigation = navigationMock(jest.fn);

  beforeEach(() => {
    mockStore = createStore(rootReducer, {
      customer: {
        data: {
          addressLine1: '123 Main Street',
          addressLine2: 'Apt 111',
          city: 'Oakland',
          state: 'CA',
          zipCode: '94501',
        },
      },
      auth: {
        credentials: {
          accessToken: 'eyJhbGciOiJSUzIYzZnpabjNoNnSJ9.eyJpc3M',
        },
      },
    });
    useNavigation.mockReturnValue(mockNavigation);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should trigger the change of screen if the address is correct', async () => {
    const { getByText, getByTestId } = render(
      <ComponentWithProviders component={UpdateAddressPage} store={mockStore} />
    );

    const addressLine1Input = getByTestId('addressLine1');
    fireEvent.changeText(addressLine1Input, '123 Main Street');

    const addressLine2Input = getByTestId('addressLine2');
    fireEvent.changeText(addressLine2Input, 'Apt 111');

    const submitButton = getByText(strings.signUp.continue);
    fireEvent.press(submitButton);

    await waitFor(() => {
      expect(mockNavigation.navigate).toBeCalledWith(NAVIGATION.settings.updateCity, {
        addressLine1: '123 Main Street',
        addressLine2: 'Apt 111',
        changed: true,
      });
    });
  });

  it('should not navigate if addressLine1 is missing', async () => {
    const { getByText, getByTestId } = render(
      <ComponentWithProviders component={UpdateAddressPage} store={mockStore} />
    );

    const addressLine2Input = getByTestId('addressLine2');
    fireEvent.changeText(addressLine2Input, 'Apt 111');

    const submitButton = getByText(strings.signUp.continue);
    fireEvent.press(submitButton);

    await waitFor(() => {
      expect(mockNavigation.navigate).toBeCalledTimes(0);
    });
  });

  it('should match snapshot', () => {
    const { toJSON } = render(
      <ComponentWithProviders component={UpdateAddressPage} store={mockStore} />
    );

    expect(toJSON()).toMatchSnapshot();
  });
});
