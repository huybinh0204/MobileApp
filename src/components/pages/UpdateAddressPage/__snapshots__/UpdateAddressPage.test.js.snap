// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateAddressPage should match snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "rgb(242, 242, 242)",
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <RNSScreenContainer
        onLayout={[Function]}
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <RNSScreen
          activityState={2}
          collapsable={false}
          forwardedRef={[Function]}
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "opacity": 1,
              }
            }
          />
          <View
            accessibilityElementsHidden={false}
            closing={false}
            gestureVelocityImpact={0.3}
            importantForAccessibility="auto"
            onClose={[Function]}
            onGestureBegin={[Function]}
            onGestureCanceled={[Function]}
            onGestureEnd={[Function]}
            onOpen={[Function]}
            onTransition={[Function]}
            pointerEvents="box-none"
            style={
              Array [
                Object {
                  "display": "flex",
                  "overflow": undefined,
                },
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
              ]
            }
            transitionSpec={
              Object {
                "close": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
                "open": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
              }
            }
          >
            <View
              collapsable={false}
              pointerEvents="box-none"
              style={
                Object {
                  "flex": 1,
                  "marginTop": 0,
                }
              }
            >
              <View
                collapsable={false}
                forwardedRef={[Function]}
                needsOffscreenAlphaCompositing={false}
                onGestureHandlerEvent={[Function]}
                onGestureHandlerStateChange={[Function]}
                style={
                  Object {
                    "flex": 1,
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                      Object {
                        "translateX": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "flex": 1,
                        "overflow": "hidden",
                      },
                      Array [
                        Object {
                          "backgroundColor": "rgb(242, 242, 242)",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                        "flexDirection": "column-reverse",
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "flex": 1,
                        }
                      }
                    >
                      <RCTSafeAreaView
                        backgroundColor={null}
                        emulateUnlessSupported={true}
                        style={
                          Array [
                            Object {
                              "backgroundColor": "#FAFAFA",
                              "flex": 1,
                            },
                            undefined,
                          ]
                        }
                        testID="UpdateAddressPage"
                      >
                        <View
                          style={
                            Array [
                              Object {
                                "alignItems": "center",
                                "height": 108.5,
                                "justifyContent": "center",
                              },
                              undefined,
                            ]
                          }
                        >
                          <View
                            accessible={true}
                            collapsable={false}
                            focusable={true}
                            onClick={[Function]}
                            onResponderGrant={[Function]}
                            onResponderMove={[Function]}
                            onResponderRelease={[Function]}
                            onResponderTerminate={[Function]}
                            onResponderTerminationRequest={[Function]}
                            onStartShouldSetResponder={[Function]}
                            style={
                              Object {
                                "left": 36,
                                "opacity": 1,
                                "position": "absolute",
                                "zIndex": 10,
                              }
                            }
                            testID="BackButton"
                          >
                            <SvgMock
                              height={45.5}
                              width={45.5}
                            />
                          </View>
                          <Text
                            allowFontScaling={false}
                            style={
                              Array [
                                Object {
                                  "color": "#1F1D25",
                                  "fontFamily": "Poppins-SemiBold",
                                  "fontSize": 29,
                                  "lineHeight": 40,
                                },
                                undefined,
                              ]
                            }
                          >
                            Home Address
                          </Text>
                        </View>
                        <View
                          onLayout={[Function]}
                          style={
                            Array [
                              Array [
                                Object {
                                  "flex": 1,
                                },
                                undefined,
                              ],
                              Object {
                                "paddingBottom": 0,
                              },
                            ]
                          }
                        >
                          <View
                            style={
                              Array [
                                Object {
                                  "flex": 1,
                                  "padding": 58,
                                },
                                undefined,
                              ]
                            }
                          >
                            <View
                              style={
                                Array [
                                  Object {
                                    "marginBottom": 87,
                                  },
                                  undefined,
                                ]
                              }
                            >
                              <View
                                style={
                                  Array [
                                    Object {
                                      "alignItems": "center",
                                      "borderRadius": 6,
                                      "borderWidth": 2,
                                      "flexDirection": "row",
                                      "width": "100%",
                                      "zIndex": 0,
                                    },
                                    Object {
                                      "borderColor": undefined,
                                    },
                                  ]
                                }
                              >
                                <BVLinearGradient
                                  colors={
                                    Array [
                                      309817028,
                                      318208316,
                                    ]
                                  }
                                  endPoint={
                                    Object {
                                      "x": 0.75,
                                      "y": 0.5,
                                    }
                                  }
                                  locations={null}
                                  startPoint={
                                    Object {
                                      "x": 0.25,
                                      "y": 0.5,
                                    }
                                  }
                                  style={
                                    Array [
                                      Object {
                                        "borderRadius": 6,
                                        "height": "100%",
                                        "position": "absolute",
                                        "width": "100%",
                                        "zIndex": 1,
                                      },
                                      Object {
                                        "opacity": undefined,
                                      },
                                    ]
                                  }
                                />
                                <Text
                                  allowFontScaling={false}
                                  onLayout={[Function]}
                                  style={
                                    Array [
                                      Object {
                                        "bottom": 0,
                                        "fontFamily": "Poppins",
                                        "fontSize": 29,
                                        "left": 25.5,
                                        "lineHeight": 40,
                                        "position": "absolute",
                                        "zIndex": 2,
                                      },
                                      Object {
                                        "color": undefined,
                                        "transform": Array [
                                          Object {
                                            "translateY": undefined,
                                          },
                                          Object {
                                            "translateX": undefined,
                                          },
                                          Object {
                                            "scale": undefined,
                                          },
                                        ],
                                      },
                                    ]
                                  }
                                >
                                  Street address
                                </Text>
                                <TextInput
                                  autoComplete="street-address"
                                  onBlur={[Function]}
                                  onChangeText={[Function]}
                                  onFocus={[Function]}
                                  style={
                                    Array [
                                      Object {
                                        "color": "#1F1D25",
                                        "flex": 1,
                                        "fontFamily": "Poppins-SemiBold",
                                        "fontSize": 29,
                                        "lineHeight": 40,
                                        "paddingBottom": 18,
                                        "paddingHorizontal": 25.5,
                                        "paddingTop": 43.5,
                                        "zIndex": 3,
                                      },
                                      undefined,
                                    ]
                                  }
                                  testID="addressLine1"
                                  textContentType="streetAddressLine1"
                                  value=""
                                />
                              </View>
                            </View>
                            <View
                              style={
                                Array [
                                  Object {
                                    "marginBottom": 87,
                                  },
                                  undefined,
                                ]
                              }
                            >
                              <View
                                style={
                                  Array [
                                    Object {
                                      "alignItems": "center",
                                      "borderRadius": 6,
                                      "borderWidth": 2,
                                      "flexDirection": "row",
                                      "width": "100%",
                                      "zIndex": 0,
                                    },
                                    Object {
                                      "borderColor": undefined,
                                    },
                                  ]
                                }
                              >
                                <BVLinearGradient
                                  colors={
                                    Array [
                                      309817028,
                                      318208316,
                                    ]
                                  }
                                  endPoint={
                                    Object {
                                      "x": 0.75,
                                      "y": 0.5,
                                    }
                                  }
                                  locations={null}
                                  startPoint={
                                    Object {
                                      "x": 0.25,
                                      "y": 0.5,
                                    }
                                  }
                                  style={
                                    Array [
                                      Object {
                                        "borderRadius": 6,
                                        "height": "100%",
                                        "position": "absolute",
                                        "width": "100%",
                                        "zIndex": 1,
                                      },
                                      Object {
                                        "opacity": undefined,
                                      },
                                    ]
                                  }
                                />
                                <Text
                                  allowFontScaling={false}
                                  onLayout={[Function]}
                                  style={
                                    Array [
                                      Object {
                                        "bottom": 0,
                                        "fontFamily": "Poppins",
                                        "fontSize": 29,
                                        "left": 25.5,
                                        "lineHeight": 40,
                                        "position": "absolute",
                                        "zIndex": 2,
                                      },
                                      Object {
                                        "color": undefined,
                                        "transform": Array [
                                          Object {
                                            "translateY": undefined,
                                          },
                                          Object {
                                            "translateX": undefined,
                                          },
                                          Object {
                                            "scale": undefined,
                                          },
                                        ],
                                      },
                                    ]
                                  }
                                >
                                  Apt/Suite (Optional)
                                </Text>
                                <TextInput
                                  onBlur={[Function]}
                                  onChangeText={[Function]}
                                  onFocus={[Function]}
                                  style={
                                    Array [
                                      Object {
                                        "color": "#1F1D25",
                                        "flex": 1,
                                        "fontFamily": "Poppins-SemiBold",
                                        "fontSize": 29,
                                        "lineHeight": 40,
                                        "paddingBottom": 18,
                                        "paddingHorizontal": 25.5,
                                        "paddingTop": 43.5,
                                        "zIndex": 3,
                                      },
                                      undefined,
                                    ]
                                  }
                                  testID="addressLine2"
                                  textContentType="streetAddressLine2"
                                  value=""
                                />
                              </View>
                            </View>
                            <View
                              style={
                                Array [
                                  Object {
                                    "alignItems": "center",
                                    "flex": 1,
                                    "justifyContent": "flex-end",
                                  },
                                  undefined,
                                ]
                              }
                            >
                              <View
                                accessibilityState={
                                  Object {
                                    "disabled": true,
                                  }
                                }
                                accessible={true}
                                collapsable={false}
                                focusable={true}
                                onClick={[Function]}
                                onResponderGrant={[Function]}
                                onResponderMove={[Function]}
                                onResponderRelease={[Function]}
                                onResponderTerminate={[Function]}
                                onResponderTerminationRequest={[Function]}
                                onStartShouldSetResponder={[Function]}
                                style={
                                  Object {
                                    "alignSelf": "stretch",
                                    "backgroundColor": "#D4D7DB",
                                    "borderRadius": 10,
                                    "opacity": 1,
                                    "paddingHorizontal": 45.5,
                                    "paddingVertical": 31,
                                  }
                                }
                                testID="signUpNameSubmitButton"
                              >
                                <Text
                                  allowFontScaling={false}
                                  disabled={true}
                                  style={
                                    Array [
                                      Object {
                                        "color": "#FFFFFF",
                                        "fontFamily": "Poppins-SemiBold",
                                        "fontSize": 25.5,
                                        "lineHeight": 35.5,
                                        "textAlign": "center",
                                      },
                                      undefined,
                                    ]
                                  }
                                >
                                  Continue
                                </Text>
                              </View>
                            </View>
                          </View>
                        </View>
                      </RCTSafeAreaView>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </RNSScreen>
      </RNSScreenContainer>
      <View
        collapsable={false}
        pointerEvents="box-none"
        style={
          Object {
            "height": 44,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "zIndex": 1,
          }
        }
      />
    </View>
  </RNCSafeAreaProvider>
</View>
`;
